src/engines/CP2K_engine.py

```diff
@@ -0,0 +1,111 @@
+"""
+Engine implementation of CP2K
+"""
+from typing import Sequence
+import numpy as np
+from cp2k_input_tools.parser import CP2KInputParser
+
+from engines.abstract_engine import AbstractEngine, ShootingResult
+
+
+class CP2KEngine(AbstractEngine):
+    def __init__(self, inputs):
+        super().__init__(inputs)
+
+        self._atoms = None
+        with open(inputs["cp2k_inputs"]) as f:
+            parser = CP2KInputParser()
+            self.cp2k_inputs = parser.parse(f)
+
+    @property
+    def atoms(self) -> Sequence[str]:
+        if self._atoms is None:
+            # TODO: How does this handle coordinates linked in a separate file?
+            # Return the first two places for each coordinate entry
+            self._atoms = [entry[0:2] for entry in self._get_coord()]
+
+        return self._atoms
+
+    def set_positions(self, positions: np.ndarray) -> None:
+        # Check positions are valid by passing to base class
+        super().set_positions(positions)
+
+        # coords stored as list of "El x y z" strings, same as CP2K .inp file
+        coords = self._get_coord()
+
+        for i in range(positions.shape[0]):
+            # Create the space separated string and append it to the atom
+            pos_str = ' '.join([str(p) for p in positions[i, :]])
+            coords[i] = f"{self.atoms[i]} {pos_str}"
+
+    def set_velocities(self, velocities: np.ndarray) -> None:
+        # Check velocities are valid by passing to base class
+        super().set_velocities(velocities)
+
+        vel = self._get_velocity()
+
+        # Assign all the velocities
+        for i in range(velocities.shape[0]):
+            for j in range(3):
+                vel[i][j] = velocities[i, j]
+
+    def validate_inputs(self, inputs: dict) -> (bool, str):
+        if "cp2k_inputs" not in inputs:
+            return False, "cp2k_inputs required for cp2k"
+
+        # Validate the CP2K input file. Parser will throw exceptions if invalid
+        # TODO: More specific error handling for .inp file
+        try:
+            with open(inputs["cp2k_inputs"]) as f:
+                parser = CP2KInputParser()
+                parser.parse(f)
+        except Exception as e:
+            return False, f"cp2k_inputs: {str(e)}"
+
+        # Otherwise let the base class validate
+        return super().validate_inputs(inputs)
+
+    async def run_shooting_point(self) -> ShootingResult:
+        pass
+
+    @property
+    def delta_t(self) -> float:
```

Comment:
Are these `delta_t` property methods intentionally empty and to be developed in later PRs?

Reply:
exactly